### PR TITLE
Add NFC support to String.normalize/2

### PR DESCRIPTION
**Pull request by endersstocker**

### CODE REVIEW

File: lib/elixir/unicode/unicode.ex [hunk 1]
```diff
@@ -482,4 +496,50 @@ defmodule String.Normalizer do
   end
 
   defp combining_class(_), do: 0
+
+  def nfd_to_nfc("", acc), do: acc
```
Comment (josevalim) [inline lib/elixir/unicode/unicode.ex]: Should those functions be private. :)

Comment (endersstocker) [inline lib/elixir/unicode/unicode.ex]: Good catch! The irony is that I added syntax support for distinguishing private functions from public ones in Sublime Text, but I never bothered to utilize that distinction in my color scheme. :laughing:

**Comment by josevalim** [comment]
@endersstocker curiosity, can you tell me how big is the String.Normalizer.beam in KB after this PR?

**Comment by endersstocker** [comment]
172

**Comment by endersstocker** [comment]
168

**Comment by endersstocker** [comment]
This implementation has to traverse the string twice. I’m working on a single-traversal implementation.

**Comment by endersstocker** [comment]
@josevalim I’ve optimized it to normalize NFC in a single pass.

File: lib/elixir/unicode/unicode.ex [hunk 2]
```diff
@@ -470,6 +480,21 @@ defmodule String.Normalizer do
...
+  defp normalize_nfc(<<cp :: utf8, rest :: binary>>, acc) when cp in 0xAC00..0xD7A3 do
+    normalize_nfc(rest, acc <> <<cp :: utf8>>)
+  end
+
+  defp normalize_nfc(binary, acc) do
+    {n, rest} = String.Graphemes.next_grapheme_size(binary)
+    part = :binary.part(binary, 0, n)
+    case n do
+      1 -> normalize_nfc(rest, acc <> part)
+      _ -> normalize_nfc(rest, acc <> compose(normalize(part, :nfd)))
```
Comment (josevalim) [inline lib/elixir/unicode/unicode.ex]: let's call `normalize_nfd(..., "")`

**Comment by josevalim** [comment]
This is great. I have added a tiny comment. Can you please address it and squashh your changes? :heart:

**Comment by endersstocker** [comment]
Good to go.

**Comment by endersstocker** [comment]
It’s good to go.

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Pull request merged**